docs/03-language-guide/14-equality.md

```diff
@@ -1,6 +1,133 @@
+
+## Collection types
+
+Wing contains six collection types: `Array`, `MutArray`, `Map`, `MutMap`, `Set`, and `MutSet`.The following rules apply to all of them:
+
+1. Two collections are equal if they have the same number of elements, and if each element in the first collection is equal to the corresponding element in the second collection. (The order of elements only matters for `Array` and `MutArray`.)
+2. The mutability of a collection does not affect its equality. In other words, a `MutArray` is equal to an `Array` with the same elements, and a `MutMap` is equal to a `Map` with the same keys and values.
+3. Only collections of the same "kind" can be equal. For example, an `Array` cannot be equal to a `Map`, and a `MutArray` cannot be equal to a `MutMap`.
+
+## Function types
+
+Two functions are equal if they are both the same function. This means that two functions that have the same code are not necessarily equal, since they may have been defined in different places.
+
+```js
+let f1 = (x: num): num => x + 1;
+let f2 = (x: num): num => x + 1;
+let f3 = f1;
+
+assert(f1 != f2);
+assert(f1 == f3);
+```
+
+## Enums
+
+Two enum values are equal if they refer to the same enum case.
+
+```js
+enum PizzaTopping {
+  CHEESE,
+  PINEAPPLE,
+}
+
+assert(PizzaTopping.CHEESE == PizzaTopping.CHEESE);
+assert(PizzaTopping.CHEESE != PizzaTopping.PINEAPPLE);
+```
+
+## Classes and interfaces
+
+Two class instances or interface-satisfying objects are equal if they are the same instance. This means that two class instances, or interface-satisfying objects that have the same data are not necessarily equal, since they may have been created in different places.
+
+```js
+class Shop {
+  hats: num;
+  init(hats: num) {
+    this.hats = hats;
+  }
+}
+
+let shop1 = new Shop(1);
+let shop2 = new Shop(1);
+let shop3 = shop1;
+
+assert(shop1 != shop2);
+assert(shop1 == shop3);
+```
+
+## `Json`
+
+Two `Json` values are equal if they contain the same structure and values. The following rules apply:
+
+1. Two `Json` values are equal if they are both `null`.
+2. Two `Json` values are equal if they are both `bool` values and are equal.
+3. Two `Json` values are equal if they are both `num` values and are equal.
+4. Two `Json` values are equal if they are both `str` values and are equal.
+5. Two `Json` values are equal if they are both `Array` values and are equal.
+6. Two `Json` values are equal if they are both `Map` values and are equal.
+
+```js
+assert(Json nil == Json nil); // TODO: https://github.com/winglang/wing/issues/1819
+assert(Json true == Json true);
+assert(Json false == Json false);
+assert(Json 1 == Json 1);
+assert(Json -0.42 == Json -0.42);
+assert(Json "foo" == Json "foo");
+assert(Json [1, 2, 3] == Json [1, 2, 3]);
+assert(Json { "foo": 1, "bar": 2 } == Json { "foo": 1, "bar": 2 });
+```
+
+## Structs
+
+Two struct values are equal if they have the same type and all of their fields are equal. The order in which fields are defined does not matter.
```

Comment:
```suggestion
Two struct values are equal if they have the same type and all of their fields are equal (based on rules of equality of their type). The order in which fields are defined does not matter.
```

Reply:
The comment about the order sees a bit misleading to me. If they are the same type they are the same type.

Reply:
I guess the question how does someone know that `Foo { a: 1, b: 2 }` is the same as `Foo { b: 2, a: 1 }`? It sounds like you're saying that's already part of the definition of a struct, in which case I'm ok taking that sentence out.